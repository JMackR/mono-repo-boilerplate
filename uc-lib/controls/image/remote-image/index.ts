export * from "./remote-image"
export * from "./remote-image.props"
