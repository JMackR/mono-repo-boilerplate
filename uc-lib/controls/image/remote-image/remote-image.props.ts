import { RemoteImageSource } from "../remote-image-source"
import { ResizeMode } from "../image-props"

export interface RemoteImageProps {
  resizeMode: ResizeMode
  source: RemoteImageSource

  width?: number | string
  height?: number | string
  aspectRatio?: number
  borderRadius?: number

  children?: React.ReactNode
  /**
   * Used to locate this view in end-to-end tests.
   */
  testID?: string
}
