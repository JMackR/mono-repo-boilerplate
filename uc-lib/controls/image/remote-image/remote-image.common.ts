import { RemoteImageProps } from "./remote-image.props"
import invariant from "invariant"

export const sanitizeAndValidateRemoteImageLayoutProps = (props: RemoteImageProps) => {
  let { width, height } = props
  const { aspectRatio } = props
  const hasWidthAndHeight = width !== undefined && height !== undefined
  const hasWidthOrHeight = width !== undefined || height !== undefined
  const hasAspectRatio = aspectRatio !== undefined
  const layoutPropsValid = hasWidthAndHeight || (hasWidthOrHeight && hasAspectRatio)

  invariant(
    layoutPropsValid === true,
    "RemoteImage's layout props are not satisfied.\n(width and height) or ((width or height) and aspectRatio) must be set",
  )

  if (hasAspectRatio) {
    if (width === undefined && typeof height === "number") {
      width = (aspectRatio as number) * (height as number)
    } else if (typeof width === "number") {
      height = (width as number) / (aspectRatio as number)
    }
  }

  return { width, height, aspectRatio }
}
