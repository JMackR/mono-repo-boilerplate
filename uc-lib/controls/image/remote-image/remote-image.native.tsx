import React, { useState, memo } from "react"
import { View } from "react-native"
import FastImage from "react-native-fast-image"
import { useColor } from "uc-lib/themes"
import { RemoteImageProps } from "./remote-image.props"
import { sanitizeAndValidateRemoteImageLayoutProps } from "./remote-image.common"
import { FC } from "react"

export const RemoteImage: FC<RemoteImageProps> = props => {
  const { width, height, aspectRatio } = sanitizeAndValidateRemoteImageLayoutProps(props)
  const { resizeMode, source, children, borderRadius, testID } = props

  const [isLoading, setIsLoading] = useState(true)

  const handleLoadEnd = () => {
    setIsLoading(false)
  }

  return (
    <View
      style={{
        justifyContent: "center",
        alignItems: "center",
        width,
        height,
        aspectRatio,
      }}
      testID={testID || "uc-lib.remote-image"}
      accessibilityLabel={testID || "uc-lib.remote-image"}
    >
      <FastImage
        source={source}
        resizeMode={resizeMode}
        onLoadEnd={handleLoadEnd}
        style={{
          width,
          height,
          aspectRatio,
          borderRadius,
        }}
      >
        {children}
      </FastImage>
      {isLoading && <LoadingView />}
    </View>
  )
}

const LoadingView: FC = memo(() => {
  const { colors } = useColor()

  return (
    <View
      style={{
        position: "absolute",
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        backgroundColor: colors.limestone,
      }}
      testID="redibs-ucl.remote-image.loading-view"
    />
  )
})
