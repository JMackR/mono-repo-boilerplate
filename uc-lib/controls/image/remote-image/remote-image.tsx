import React, { useState, useEffect, memo } from "react"
import { useColor } from "uc-lib/themes"
import { RemoteImageProps } from "./remote-image.props"
import { sanitizeAndValidateRemoteImageLayoutProps } from "./remote-image.common"
import { StyleSheet, css } from "aphrodite/no-important"
import _ from "lodash"
export const RemoteImage = memo((props: RemoteImageProps) => {
  const { width, height, aspectRatio } = sanitizeAndValidateRemoteImageLayoutProps(props)
  const aspectRatioToPercentageMultiplier = 100
  const paddingTopToMaintainAspectRatio =
    (aspectRatio && aspectRatioToPercentageMultiplier / aspectRatio + "%") ||
    `${(_.toNumber(width) / _.toNumber(height)) * 100}%`

  const { resizeMode, source, children, borderRadius } = props
  const objectFit = objectFitForResizeMode(resizeMode)

  const [isLoading, setIsLoading] = useState(true)
  const { colors } = useColor()
  let isMounted = false
  useEffect(() => {
    isMounted = true
  }, [])
  const handleLoadEnd = () => {
    setIsLoading(false)
  }
  const STYLES = StyleSheet.create({
    loading: {
      position: "absolute",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      backgroundColor: colors.limestone,
    },
    container: {
      position: "relative",
      justifyContent: "center",
      alignItems: "center",
      display: "block",
      width,
      height: 0,
      borderRadius,
      paddingTop: paddingTopToMaintainAspectRatio,
      overflow: "hidden",
      minWidth: width,
    },
    image: {
      position: "absolute",
      maxWidth: "100%",
      top: 0,
      bottom: 0,
      height: "100%",
      width: "100%",
      objectFit,
      display: "block",
    },
    children: {
      position: "absolute",
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      zIndex: 2,
    },
  })

  const getLoadingView = () => {
    return <div className={css(STYLES.loading)} />
  }

  return (
    <picture className={css(STYLES.container)}>
      <source srcSet={source.uri} />
      <img src={source.uri} onLoad={handleLoadEnd} className={css(STYLES.image)} />
      <div className={css(STYLES.children)}>{children}</div>
      {isMounted && isLoading && getLoadingView()}
    </picture>
  )
})

const objectFitForResizeMode = (resizeMode: string) => {
  let objectFit: "cover" | "contain" | "none" = "none"
  if (resizeMode === "cover") {
    objectFit = "cover"
  } else if (resizeMode === "contain") {
    objectFit = "contain"
  }
  return objectFit
}
