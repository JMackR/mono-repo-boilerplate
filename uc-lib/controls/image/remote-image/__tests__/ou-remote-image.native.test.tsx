import React from "react"
import { Text } from "react-native"
import { render, shallow } from "react-native-testing-library"
import { RemoteImage } from "../remote-image"

describe("RemoteImage Snapshot Tests", () => {
  const photoURL =
    "https://media-exp1.licdn.com/dms/image/C4E03AQHXHzRJLFryWw/profile-displayphoto-shrink_200_200/0?e=1585785600&v=beta&t=36ZZ3awsowdk_mAFr9oTQ5IS9moXDqNP9jRx_qR1hJU"
  const width = 100
  const height = 150
  const borderRadius = 4
  const aspectRatio = 1.2

  test("RemoteImage renders width/height/resizeMode/borderRadius", () => {
    const tree = render(
      <RemoteImage
        width={width}
        height={height}
        resizeMode="center"
        borderRadius={borderRadius}
        source={{ uri: photoURL }}
      />,
    )
    expect(tree).toMatchSnapshot()
  })

  test("RemoteImage renders width/height/resizeMode/borderRadius and children", () => {
    const tree = render(
      <RemoteImage width={width} height={height} resizeMode="center" borderRadius={borderRadius} source={{ uri: photoURL }}>
        <Text>Hype!</Text>
      </RemoteImage>,
    )
    expect(tree).toMatchSnapshot()
  })

  test("RemoteImage renders width/aspectRatio/resizeMode", () => {
    const tree = render(
      <RemoteImage
        width={width}
        aspectRatio={aspectRatio}
        resizeMode="contain"
        borderRadius={0}
        source={{ uri: photoURL }}
      />,
    )
    expect(tree).toMatchSnapshot()
  })

  test("RemoteImage renders height/aspectRatio/resizeMode", () => {
    const tree = render(
      <RemoteImage height={height} aspectRatio={aspectRatio} resizeMode="cover" source={{ uri: photoURL }} />,
    )
    expect(tree).toMatchSnapshot()
  })

  test("RemoteImage does not render with only height", () => {
    expect(() => shallow(<RemoteImage height={height} resizeMode="cover" source={{ uri: photoURL }} />)).toThrowError()
  })

  test("RemoteImage does not render with only width", () => {
    expect(() => shallow(<RemoteImage width={width} resizeMode="cover" source={{ uri: photoURL }} />)).toThrowError()
  })

  test("RemoteImage does not render with only aspectRatio", () => {
    expect(() =>
      shallow(<RemoteImage aspectRatio={aspectRatio} resizeMode="cover" source={{ uri: photoURL }} />),
    ).toThrowError()
  })
})
