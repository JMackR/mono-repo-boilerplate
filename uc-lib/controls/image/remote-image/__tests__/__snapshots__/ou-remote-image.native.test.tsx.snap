// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteImage Snapshot Tests RemoteImage renders height/aspectRatio/resizeMode 1`] = `
<View
  accessibilityLabel="redibs-ucl.remote-image"
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1.2,
      "height": 150,
      "justifyContent": "center",
      "width": 180,
    }
  }
  testID="redibs-ucl.remote-image"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "aspectRatio": 1.2,
          "borderRadius": undefined,
          "height": 150,
          "width": 180,
        },
      ]
    }
  >
    <FastImageView
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://media-exp1.licdn.com/dms/image/C4E03AQHXHzRJLFryWw/profile-displayphoto-shrink_200_200/0?e=1585785600&v=beta&t=36ZZ3awsowdk_mAFr9oTQ5IS9moXDqNP9jRx_qR1hJU",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="redibs-ucl.remote-image.loading-view"
  />
</View>
`;

exports[`RemoteImage Snapshot Tests RemoteImage renders width/aspectRatio/resizeMode 1`] = `
<View
  accessibilityLabel="redibs-ucl.remote-image"
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1.2,
      "height": 83.33333333333334,
      "justifyContent": "center",
      "width": 100,
    }
  }
  testID="redibs-ucl.remote-image"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "aspectRatio": 1.2,
          "borderRadius": 0,
          "height": 83.33333333333334,
          "width": 100,
        },
      ]
    }
  >
    <FastImageView
      onFastImageLoadEnd={[Function]}
      resizeMode="contain"
      source={
        Object {
          "uri": "https://media-exp1.licdn.com/dms/image/C4E03AQHXHzRJLFryWw/profile-displayphoto-shrink_200_200/0?e=1585785600&v=beta&t=36ZZ3awsowdk_mAFr9oTQ5IS9moXDqNP9jRx_qR1hJU",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="redibs-ucl.remote-image.loading-view"
  />
</View>
`;

exports[`RemoteImage Snapshot Tests RemoteImage renders width/height/resizeMode/borderRadius 1`] = `
<View
  accessibilityLabel="redibs-ucl.remote-image"
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": undefined,
      "height": 150,
      "justifyContent": "center",
      "width": 100,
    }
  }
  testID="redibs-ucl.remote-image"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "aspectRatio": undefined,
          "borderRadius": 4,
          "height": 150,
          "width": 100,
        },
      ]
    }
  >
    <FastImageView
      onFastImageLoadEnd={[Function]}
      resizeMode="center"
      source={
        Object {
          "uri": "https://media-exp1.licdn.com/dms/image/C4E03AQHXHzRJLFryWw/profile-displayphoto-shrink_200_200/0?e=1585785600&v=beta&t=36ZZ3awsowdk_mAFr9oTQ5IS9moXDqNP9jRx_qR1hJU",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="redibs-ucl.remote-image.loading-view"
  />
</View>
`;

exports[`RemoteImage Snapshot Tests RemoteImage renders width/height/resizeMode/borderRadius and children 1`] = `
<View
  accessibilityLabel="redibs-ucl.remote-image"
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": undefined,
      "height": 150,
      "justifyContent": "center",
      "width": 100,
    }
  }
  testID="redibs-ucl.remote-image"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "aspectRatio": undefined,
          "borderRadius": 4,
          "height": 150,
          "width": 100,
        },
      ]
    }
  >
    <FastImageView
      onFastImageLoadEnd={[Function]}
      resizeMode="center"
      source={
        Object {
          "uri": "https://media-exp1.licdn.com/dms/image/C4E03AQHXHzRJLFryWw/profile-displayphoto-shrink_200_200/0?e=1585785600&v=beta&t=36ZZ3awsowdk_mAFr9oTQ5IS9moXDqNP9jRx_qR1hJU",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <Text>
      Hype!
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="redibs-ucl.remote-image.loading-view"
  />
</View>
`;
